src/NUnitFramework/framework/Interfaces/TNode.cs

```diff
@@ -318,17 +319,64 @@ private static NodeList ApplySelection(NodeList nodeList, string xpath)
                 : resultNodes;
         }
 
-        private static readonly Regex InvalidXmlCharactersRegex = new Regex("[^\u0009\u000a\u000d\u0020-\ufffd]|([\ud800-\udbff](?![\udc00-\udfff]))|((?<![\ud800-\udbff])[\udc00-\udfff])", RegexOptions.Compiled);
         private static string EscapeInvalidXmlCharacters(string str)
         {
             if (str == null) return null;
 
             // Based on the XML spec http://www.w3.org/TR/xml/#charsets
-            // For detailed explanation of the regex see http://mnaoumov.wordpress.com/2014/06/15/escaping-invalid-xml-unicode-characters/
 
-            return InvalidXmlCharactersRegex.Replace(str, match => CharToUnicodeSequence(match.Value[0]));
-        }
+            StringBuilder builder = null;
+            for (int i = 0; i < str.Length; i++)
+            {
+                char c = str[i];
+                if(c > 0x20 && c < 0x7F)
+                {
+                    // ASCII characters - break quickly for these
+                    if (builder != null)
+                        builder.Append(c);
+                }
+                else if (!(0x0 <= c && c <= 0x8) &&
+                    c != 0xB &&
+                    c != 0xC &&
+                    !(0xE <= c && c <= 0x1F) &&
+                    !(0x7F <= c && c <= 0x84) &&
+                    !(0x86 <= c && c <= 0x9F) &&
+                    !(0xD800 <= c && c <= 0xDFFF) &&
+                    c != 0xFFFE &&
+                    c != 0xFFFF)
```

Comment:
Would it be possible to add some comments to this part of the if-statement? Comparing with https://www.w3.org/TR/xml/#charsets I think that I get some of the cases, but every time I need to look at this code I'll need to look up what e.g. `0x85` is, or why the range `0x86 <= c && c <= 0x9F` is special

Reply:
Hopefully these comments should be enough for you - let me know if there's anything else you'd like.